test/testutilsversion.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2022  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include <QTest>
+
+#include "base/global.h"
+#include "base/utils/version.h"
+
+class TestUtilsVersion final : public QObject
+{
+    Q_OBJECT
+    Q_DISABLE_COPY_MOVE(TestUtilsVersion)
+
+public:
+    TestUtilsVersion() = default;
+
+private slots:
+    void testConstructors() const
+    {
+        using TwoDigits = Utils::Version<unsigned char, 2, 1>;
+        QCOMPARE(TwoDigits(), TwoDigits());
+        QCOMPARE(TwoDigits(), TwoDigits(0));
```

Comment:
Ok, let's assume the default constructor is allowed for some reason. But it creates "invalid" version, doesn't it? Then I don't think invalid version can be equal to any valid one.

Reply:
>But it creates "invalid" version, doesn't it? 

Yes, it initialize to the type (`unsigned char`) default value. But at the same time `Version::isValid` would say `TwoDigits(0)` is invalid too.
This is another existing problem...

Anyway I should avoid comparing invalid values to avoid confusion.

Reply:
Don't you think it would be right to fix all the known flaws of `Version` class before adding tests for it? If we just avoid touching on these shortcomings in the tests, then what is the use of them?

As far as I remember, this class raised questions to its logic from scratch, and it still does not look reliable and consistent. Maybe it's time to finally fix it, since you're thinking of adding it to testing?

Reply:
>Don't you think it would be right to fix all the known flaws of Version class before adding tests for it? If we just avoid touching on these shortcomings in the tests, then what is the use of them?

This process can be iterative. I have solved some problems in this PR and the rest can be solved in another.

>Maybe it's time to finally fix it, since you're thinking of adding it to testing?

There are problems everywhere, this is just the least significant portion of it that I choose to submit first. Unit testing is useful in uncover usability issues like this. However you can not expect a Class to be perfect before adding some testing otherwise the requirement would be too high and will block everything.

As for the problem. (In a follow up PR) I would utilize `std::optional` in Version internal to handle the validness issue. Or do you have better ideas or wish to tackle the problem yourself?

Reply:
> do you have better ideas

No.
This is the first thing that comes to mind, and I have not considered its implementation in more detail. The main thing is that Version eventually has the correct interface and behavior.

> wish to tackle the problem yourself?

No. Currently it doesn't bother me in practice.